### ISSUE TITLE
Make the terminal experience less latency-sensitive

### ISSUE THREAD

**Issue by ishitatsuyuki** [JanitorTechnology/janitor#286]
If you have a ping of over 300ms, the terminal feels much less like at home. There's a few thing we can do to improve that:

- Preinstall mosh to allow lag and cellular friendly direct SSH connections.
- Enable the `c9.ide.terminal/predict_echo` plugin, which writes out the characters as you type so it feels less laggy (similar to mosh's local echo feature).

Note that @jankeromnes mentioned on IRC that the c9 plugin had some issues, but I don't know what exactly it is.

**Comment by jankeromnes** [comment]
Thanks a lot for explaining the problem, and for making very helpful suggestions! As a side-note, I believe this issue would be better-suited to our [Janitor Cloud9 plugin](https://github.com/JanitorTechnology/c9.ide.janitorconfig) (and we can also install additional Cloud9 plugins in all images by cloning them [into the ubuntu-dev image](https://github.com/JanitorTechnology/dockerfiles/blob/1a38902a5cf6a33700e8c44883769add51535903/ubuntu-dev/ubuntu-dev.dockerfile#L215-L225)).

> Preinstall mosh to allow lag and cellular friendly direct SSH connections.

I believe that Cloud9's terminal uses tmux. Could mosh work on top of that? @EtienneWan would you have any insights into how doable this would be?

> Enable the `c9.ide.terminal/predict_echo` plugin, which writes out the characters as you type so it feels less laggy (similar to mosh's local echo feature)

Aha, that sounds interesting. It would be worth investigating if this plugin is already installed and/or enabled in our Cloud9 SDK checkout. If not, we should probably try installing and/or enabling it, to see if this improves the situation.

> Note that @jankeromnes mentioned on IRC that the c9 plugin had some issues, but I don't know what exactly it is.

From experience when using the Cloud9 Terminal on a bad connection, I've noticed that often it lets me type several characters in advance fluidly, but then when you type `Enter` it actually has to wait for the SSH connection to catch up. Also, sometimes the "type ahead" feature doesn't even work reliably (e.g. after you typed two words, it may decide it needs to catch up now, and the output freezes).

I have no idea if this is `c9.ide.terminal/predict_echo` plugin or something else, but if we have a way to improve the Terminal experience, I'd love to try it!

**Comment by ishitatsuyuki** [comment]
> I believe that Cloud9's terminal uses tmux. Could mosh work on top of that?

Yes, mosh even works best with tmux as it covers the scrolling functionality which mosh doesn't have by design.

> From experience when using the Cloud9 Terminal on a bad connection, I've noticed that often it lets me type several characters in advance fluidly, but then when you type Enter it actually has to wait for the SSH connection to catch up. Also, sometimes the "type ahead" feature doesn't even work reliably (e.g. after you typed two words, it may decide it needs to catch up now, and the output freezes for several seconds before you even finished typing your line).

I think that's unfortunately how it *should* be implemented. I think it will still improve the situation.
